### PR TITLE
Refactor part of Iceberg code to support more Iceberg Catalogs

### LINKED ISSUE #9657: [Enhancement] Support more type of Iceberg catalogs
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Support more types of Iceberg catalogs. Iceberg supports many catalogs (such as HiveCatalog, HadoopCatalog, JDBCCatalog, GlueCatalog, and more). Currently, we only support HiveCatalog in Doris. This should be improved and also we should support set catalog properties, such as `clients` (HMS clients) for HiveCatalog and similar properties. 

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by ConeyLiu**
# Proposed changes

Issue Number: close #9657

## Problem Summary:

This PR refactors part of the Iceberg code to support more Iceberg Catalogs. And also it supports set more catalog specified properties.

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (Yes)
3. Has document been added or modified: (Yes)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergCatalogMgr.java [hunk 1]
```diff
@@ -50,83 +49,62 @@ public class IcebergCatalogMgr {
 
     // hive metastore uri -> iceberg catalog
     // used to cache iceberg catalogs
-    private static final ConcurrentHashMap<String, IcebergCatalog> metastoreUriToCatalog = new ConcurrentHashMap();
+    private static final ConcurrentHashMap<IcebergProperty, IcebergCatalog> propertiesToCatalog = new ConcurrentHashMap();
```
Comment (qidaye) [inline fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergCatalogMgr.java:52]: It's better to use `IcebergProperty.extraProperties`, such as `uri`/`warehouse`, as the cache key. 
`IcebergProperty` is bound to the database and table, not a generic cache.

File: fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergCatalogImpl.java [hunk 2]
```diff
@@ -14,54 +14,48 @@
...
+        Map<String, String> properties = Maps.newHashMap(icebergProperty.getExtraProperties());
+        properties.put(CatalogUtil.ICEBERG_CATALOG_TYPE, icebergProperty.getCatalogType());
+        // Hadoop configuration
         Configuration conf = new Configuration();
-        hiveCatalog.setConf(conf);
-        // initialize hive catalog
-        Map<String, String> catalogProperties = new HashMap<>();
-        catalogProperties.put("uri", icebergProperty.getHiveMetastoreUris());
-        hiveCatalog.initialize("hive", catalogProperties);
+        this.icebergCatalog = CatalogUtil.buildIcebergCatalog(icebergProperty.getCatalogType(), properties, conf);
```
Comment (qidaye) [inline fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergCatalogImpl.java:47]: **Note**: `CatalogUtil.buildIcebergCatalog` only supports `HiveCatalog` and `HadoopCatalog` for now.
How do you plan to support other catalog types?

File: fe/fe-core/src/main/java/org/apache/doris/planner/IcebergScanNode.java [hunk 3]
```diff
@@ -99,9 +98,6 @@ public String getNodeExplainString(String prefix, TExplainLevel detailLevel) {
         StringBuilder output = new StringBuilder();
         if (!isLoad()) {
             output.append(prefix).append("TABLE: ").append(icebergTable.getName()).append("\n");
-            output.append(prefix).append("PATH: ")
```
Comment (qidaye) [inline fe/fe-core/src/main/java/org/apache/doris/planner/IcebergScanNode.java:102]: Uri or warehouse information should be added here.

File: fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergCatalogMgr.java (hunk 1 above)
Comment (ConeyLiu) [reply fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergCatalogMgr.java]: Updated

File: fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergCatalogImpl.java (hunk 2 above)
Comment (ConeyLiu) [reply fe/fe-core/src/main/java/org/apache/doris/external/iceberg/IcebergCatalogImpl.java]: We could use the `iceberg.catalog.catalog-impl` property to load other catalogs with the full catalog name.

File: fe/fe-core/src/main/java/org/apache/doris/planner/IcebergScanNode.java (hunk 3 above)
Comment (ConeyLiu) [reply fe/fe-core/src/main/java/org/apache/doris/planner/IcebergScanNode.java:102]: Added

**Comment by ConeyLiu** [comment]
@qidaye, thanks for the review. Comments have been addressed. Please take another look when you are free. Thanks again.

File: docs/en/ecosystem/external-table/iceberg-of-doris.md [hunk 4]
```diff
@@ -135,10 +167,15 @@ You can also create an Iceberg table by explicitly specifying the column definit
     - The order of the columns needs to be consistent with the Iceberg table
 - ENGINE needs to be specified as ICEBERG
 - PROPERTIES property.
-    - `iceberg.hive.metastore.uris`: Hive Metastore service address
     - `iceberg.database`: the name of the database to which Iceberg is mounted
     - `iceberg.table`: the name of the table to which Iceberg is mounted, not required when mounting Iceberg database.
-    - `iceberg.catalog.type`: the catalog method used in Iceberg, the default is `HIVE_CATALOG`, currently only this method is supported, more Iceberg catalog access methods will be supported in the future.
+    - `iceberg.catalog.type`: the catalog type used in Iceberg, the default is `HIVE` (simple name of `org.apache.iceberg.hive.HiveCatalog`).
+    - `iceberg.catalog.catalog-impl`: you could also use the full catalog name with this property. You should not use both `iceberg.catalog.type` and `iceberg.catalog.catalog-impl`.
```
Comment (qidaye) [inline docs/en/ecosystem/external-table/iceberg-of-doris.md:173]: The user does not need to care about the specific implementation classes of the catalog already supported in iceberg, which would make it more difficult to use. Such as `JDBC`/`Glue` and so on.

User only needs to set the catalog type and the specific implementation class Doris should encapsulate it.

Wouldn't it be better if the user only needs to set the implementation class information when using a custom catalog?

File: docs/en/ecosystem/external-table/iceberg-of-doris.md [hunk 5]
```diff
@@ -135,10 +167,15 @@ You can also create an Iceberg table by explicitly specifying the column definit
     - The order of the columns needs to be consistent with the Iceberg table
 - ENGINE needs to be specified as ICEBERG
 - PROPERTIES property.
-    - `iceberg.hive.metastore.uris`: Hive Metastore service address
     - `iceberg.database`: the name of the database to which Iceberg is mounted
     - `iceberg.table`: the name of the table to which Iceberg is mounted, not required when mounting Iceberg database.
-    - `iceberg.catalog.type`: the catalog method used in Iceberg, the default is `HIVE_CATALOG`, currently only this method is supported, more Iceberg catalog access methods will be supported in the future.
+    - `iceberg.catalog.type`: the catalog type used in Iceberg, the default is `HIVE` (simple name of `org.apache.iceberg.hive.HiveCatalog`).
```
Comment (qidaye) [inline docs/en/ecosystem/external-table/iceberg-of-doris.md:172]: `type` only supports `HIVE` and `HADOOP`.
Other types of catalog should use `catalog-impl` config. We should make it clear for users.

File: docs/en/ecosystem/external-table/iceberg-of-doris.md (hunk 4 above)
Comment (ConeyLiu) [reply docs/en/ecosystem/external-table/iceberg-of-doris.md:173]: Do you mean always set the full class name?  Would it be inconvenient for users who use the `HIVE` or `Hadoop`?

File: docs/en/ecosystem/external-table/iceberg-of-doris.md (hunk 5 above)
Comment (ConeyLiu) [reply docs/en/ecosystem/external-table/iceberg-of-doris.md:172]: OK, will update it

File: docs/en/ecosystem/external-table/iceberg-of-doris.md (hunk 4 above)
Comment (qidaye) [reply docs/en/ecosystem/external-table/iceberg-of-doris.md:173]: > Do you mean always set the full class name? Would it be inconvenient for users who use the `HIVE` or `Hadoop`?

What I mean is that the catalog types already supported by iceberg can be set without setting the full class name, just set the type.
For custom catalog types, only the full class name needs to be set

File: docs/en/ecosystem/external-table/iceberg-of-doris.md (hunk 5 above)
Comment (ConeyLiu) [reply docs/en/ecosystem/external-table/iceberg-of-doris.md]: Updated

File: docs/en/ecosystem/external-table/iceberg-of-doris.md (hunk 4 above)
Comment (ConeyLiu) [reply docs/en/ecosystem/external-table/iceberg-of-doris.md]: Updated

File: fe/fe-core/src/main/java/org/apache/doris/catalog/IcebergProperty.java [hunk 6]
```diff
@@ -17,33 +17,73 @@
...
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+
+import java.util.Locale;
 import java.util.Map;
+import java.util.Objects;
+
+import static org.apache.iceberg.CatalogUtil.ICEBERG_CATALOG_HADOOP;
+import static org.apache.iceberg.CatalogUtil.ICEBERG_CATALOG_HIVE;
```
Comment (qidaye) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/IcebergProperty.java:31]: Do not import static member.

File: docs/en/ecosystem/external-table/iceberg-of-doris.md [hunk 7]
```diff
@@ -135,10 +167,14 @@ You can also create an Iceberg table by explicitly specifying the column definit
     - The order of the columns needs to be consistent with the Iceberg table
 - ENGINE needs to be specified as ICEBERG
 - PROPERTIES property.
-    - `iceberg.hive.metastore.uris`: Hive Metastore service address
     - `iceberg.database`: the name of the database to which Iceberg is mounted
     - `iceberg.table`: the name of the table to which Iceberg is mounted, not required when mounting Iceberg database.
-    - `iceberg.catalog.type`: the catalog method used in Iceberg, the default is `HIVE_CATALOG`, currently only this method is supported, more Iceberg catalog access methods will be supported in the future.
+    - `iceberg.catalog.type`: the catalog type used in Iceberg, the default is `HIVE` (simple name of `org.apache.iceberg.hive.HiveCatalog`). Iceberg built-in supports two catalogs with a simple name: `HIVE` and `HADOOP`. You should use the full class name for other implementations.
```
Comment (qidaye) [inline docs/en/ecosystem/external-table/iceberg-of-doris.md:172]: How about add more built-in catalogs, such as JDBC, Glue catalog, for user convenience ?

User only need set full class name when using a custom catalog.

**Pull request closed without merging**